modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache

```diff
@@ -622,6 +622,46 @@ def deserialize_primitive(data, klass, path_to_item):
         )
 
 
+def get_discriminator_class(model_class, model_data, cls_visited):
+    """Returns the child class specified by the discriminator.
+
+    Args:
+        model_class (OpenApiModel): the model class
+        model_data (list/dict): data to instantiate the model.
+        cls_visited (set): list of model classes that have been visited.
+            Used to determine the discriminator class without
+            visiting circular references indefinitely.
+
+    Returns:
+        used_model_class (class/None): the chosen child class that will be used
+            to deserialize the data, for example dog.Dog.
+            If a class is not found, None is returned.
+    """
+
+    cls_visited.add(model_class)
+    discriminator = model_class.discriminator
+    discr_propertyname_py = list(discriminator.keys())[0]
+    discr_propertyname_js = model_class.attribute_map[discr_propertyname_py]
+    if discr_propertyname_js in model_data:
+        discr_value = model_data[discr_propertyname_js]
+    elif discr_propertyname_py in model_data:
+        discr_value = model_data[discr_propertyname_py]
+    else:
+        # The input data does not contain the discriminator property.
+        # The caller will raise an exception.
+        return None
+    class_name_to_discr_class = discriminator[discr_propertyname_py]
+    used_model_class = class_name_to_discr_class.get(discr_value)
+    if used_model_class is None:
```

Comment:
Why are we looping over these values?
Do we want to loop over the ancestors here?
If we have:
allOf:
- Animal
oneOf:
- Cat
- Dog
we are looking at the discriminators in Cat and Dog.
Don't we instead want to look at the discriminator in Animal? Can you add comments here describing these different cases?

Reply:
> Why are we looping over these values?
> Do we want to loop over the ancestors here?

What is your exact definition of ancestor? Is it the allOf children?

> If we have:
> allOf:
> 
> * Animal
> 
> oneOf:
> 
> * Cat
> * Dog
> 
> we are looking at the discriminators in Cat and Dog.
> Don't we instead want to look at the discriminator in Animal? 

Sure yes in that example, but what if instead the schema is:
```yaml
oneOf:
- Animal
- Plant
```

In that case there are potentially many subclasses of animal and plant, so we would need to traverse through oneOf. Same thing for anyOf.

> Can you add comments here describing what we are doing?
> `# if we didn't find a discriminated class it may exist in an ancestor`
> Whatever you are adding, can you add a sample schema and test that verifies that this feature is working?
> Otherwise if it breaks in the future, we will have no way of seeing that it has broken.
> 
> For example [ParentPet](https://github.com/OpenAPITools/openapi-generator/blob/master/samples/client/petstore/python-experimental/petstore_api/models/parent_pet.py) only discriminates to
> 
> * ChildCat
> * ChildDog
> * ChildLizard
> 
> But it allOf includes [GrandparentAnimal](https://github.com/OpenAPITools/openapi-generator/blob/30ee3f343eb681514ce5b33438de3b8f4fb1f50b/samples/client/petstore/python-experimental/petstore_api/models/grandparent_animal.py#L57).
> which includes those AND ParentPet.
> So if we deserialize a payload:
> {petType: "ParentPet"} where data_type = parent_pet.ParentPet, we should travel back through its GrandparentAnimal ancestor to find the correct class to deserialize to, as long as the discriminator is the same.
> 
> Should we look in the current discriminator, then look in descendants, then look in ancestors?
> It looks like yes per https://github.com/OpenAPITools/openapi-generator/pull/4913/files

Reply:
We are agreed that we should loop over descendants then ancestors.
Descendants are the oneOf/anyOf children and are present in the class_name_to_discr_class.
Ancestors are in cls._composed_schemas['allOf']
Can you:
- add comments describing where we are looping over ancestors and descentents
- add code to loop over ancestors
- add a test case that demonstrates deserializing a descendent
- add a test case that demonstrates deserializing an ancestor. You should be able to do this with the ParentPet class deserializing a ParentPet
- update this function to not pass in model_data, instead pass in the value from the payload corresponding to the discriminator.propertyName. This lets us cache the results of function calls in the next item.
- cache the results of the function cals by adding a lru_cache decorator around the get_discriminator_class function. I am fine with us only adding this for python3. You can do this with pseudocode:
```
if six.PY3:
  get_discriminator_class = lru_cache(get_discriminator_class)
```

Reply:
I will not do 6 in this PR. This increases the scope of the PR and should be handled in a different PR.

Reply:
Thanks for letting me know about point 6. I can definitely add 6 in a different PR.